### DISCUSSION TITLE
How to decorate an existing service with a stacked decorator?

### DISCUSSION THREAD

**Discussion by SuperDuper3** [symfony/symfony#48418] [Q&A] [answered]
[These docs](https://symfony.com/doc/current/service_container/service_decoration.html#stacking-decorators) show how to create stacking decorators.

However, if I do:

```php
 $services->set(UserContextBuilder::class)
        ->decorate('api_platform.serializer.context_builder', priority: 10);

$services->set(ProfileContextBuilder::class)
        ->decorate('api_platform.serializer.context_builder', priority: 20);

// more decorators follow
 ```

I can still access the original `api_platform.serializer.context_builder`, which is decorated accordingly.

But if I try to do it with a stacking decorator:

```php
$services->stack(
        'decorated_context_builder',
        [
            inline_service(UserContextBuilder::class),
            inline_service(ProfileContextBuilder::class),
            service('api_platform.serializer.context_builder'),
        ]
 );
```

Now `api_platform.serializer.context_builder` remains _un_-decorated, and I would need to use `decorated_context_builder` to go through my newly created stack. Which is no good, since my dependencies are using `api_platform.serializer.context_builder` and know nothing about my newly created "stack".

How can I get the convenience of a stacking decorator, and at the same time use it to decorate an existing third party service definition?

**Comment by stof** [comment, marked as answer]
You cannot do that. Services stacks are a different way to configure services, which are not compatible with decorating other services.

**Reply by SuperDuper3** [reply]
Oh well, that sucks a bit. 

Thanks anyhow. 👍

**Reply by ahundiak** [reply]
In which case the docs should probably be tweaked slightly.  The docs show stacking services using the priority approach followed by the stacking approach and states `The result will be the same as in the previous section:` And of course they are not.

**Reply by SuperDuper3** [reply]
Yes, that was my feeling as well. The docs seem to imply, IMO, that a stacking decorator is simply a more convenient syntax for decoration, but it's not really compatible with the "regular" use.

**Reply by lyrixx** [reply]
Can someone submit a PR, or at least open an issue? thanks

**Comment by nicolas-grekas** [comment]
I had this capability on my TODO, here is the PR: https://github.com/symfony/symfony/pull/63590
